### PR TITLE
ShardingSphere-JDBC 5.0.0-RC1-SNAPSHOT: throw StringIndexOutOfBoundsException when config the encrypt rule and use the sql join

### LINKED ISSUE #13057: ShardingSphere-JDBC 5.0.0-RC1-SNAPSHOT: throw StringIndexOutOfBoundsException when config the encrypt rule and use the sql join
## Bug Report

when use sql join with two tables ,and the tables have the encrypt rule ,it will throw the StringIndexOutOfBoundsException
but with no encrypt rule for the tables, it run well .

the sql is:
```java
Query query = ss.createQuery("select a from Doctor a,Doctor1 b where a.id=b.doctorId and a.mobile=:mobile)");
                query.setParameter("mobile", mobile);
                List list = query.list();
```
the encrypt rule:

```xml
	<encrypt:rule id="encryptRule" query-with-cipher-column="false">
	<encrypt:table name="doctor">
		<encrypt:column logic-column="mobile" plain-column="${plain.column:mobile}" cipher-column="mobile_cipher" encrypt-algorithm-ref="sm4_encryptor" />
		<encrypt:column logic-column="idcard" plain-column="${plain.column:idcard}" cipher-column="idcard_cipher" encrypt-algorithm-ref="sm4_encryptor" />
	</encrypt:table>
	<encrypt:table name="doctor1">
			<encrypt:column logic-column="mobile" plain-column="Mobile" cipher-column="mobile_cipher" encrypt-algorithm-ref="sm4_encryptor" />
		</encrypt:table>
	</encrypt:rule>

	<shardingsphere:data-source id="encryptDataSource" data-source-names="dataSource"  rule-refs="encryptRule">
		<props>
			<prop key="sql-show">true</prop>
		</props>
	</shardingsphere:data-source>
```

### Which version of ShardingSphere did you use?
the master branch ,5.0.0-RC1-SNAPSHOT
### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-JDBC 
### Actual behavior
```java
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -20
	at java.lang.String.substring(String.java:1967)
	at org.apache.shardingsphere.infra.rewrite.sql.impl.AbstractSQLBuilder.getConjunctionText(AbstractSQLBuilder.java:64)
	at org.apache.shardingsphere.infra.rewrite.sql.impl.AbstractSQLBuilder.toSQL(AbstractSQLBuilder.java:47)
	at org.apache.shardingsphere.infra.rewrite.engine.RouteSQLRewriteEngine.rewrite(RouteSQLRewriteEngine.java:52)
	at org.apache.shardingsphere.infra.rewrite.SQLRewriteEntry.rewrite(SQLRewriteEntry.java:71)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.rewrite(KernelProcessor.java:59)
	at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:47)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.createExecutionContext(ShardingSpherePreparedStatement.java:378)
	at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.executeQuery(ShardingSpherePreparedStatement.java:180)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:70)
	at org.hibernate.loader.Loader.getResultSet(Loader.java:2117)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1900)
	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1876)
	at org.hibernate.loader.Loader.doQuery(Loader.java:919)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:336)
	at org.hibernate.loader.Loader.doList(Loader.java:2617)
	at org.hibernate.loader.Loader.doList(Loader.java:2600)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2429)
	at org.hibernate.loader.Loader.list(Loader.java:2424)
	at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:501)
	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:371)
	at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:216)
	at org.hibernate.internal.StatelessSessionImpl.list(StatelessSessionImpl.java:566)
	at org.hibernate.internal.QueryImpl.list(QueryImpl.java:87)
```
### Reason analyze (If you can)

it works fine with 5.0.0-beta, but 5.0.0-RC1-SNAPSHOT not work fine .

with 5.0.0-RC1-SNAPSHOT,  
Scene 1:
- query-with-cipher-column="false" 
- i found that if one the of tables was not in encrypt rule, it works fine 。
- all of the two tables config the encrypt rule, it   works not  fine 。
Scene 2:
- query-with-cipher-column="true" 
- all of the two tables config the encrypt rule, it  works fine 。

### PR DESCRIPTION
**Pull request by LeeGuoPing**
Fixes #13057.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

**Comment by LeeGuoPing** [comment]
> Hi @LeeGuoPing, can you take a look for ci exception?

@strongduanmu resolved

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptProjectionTokenGenerator.java [hunk 1]
```diff
@@ -104,6 +105,44 @@ protected boolean isGenerateSQLTokenForEncrypt(final SQLStatementContext sqlStat
...
+    
+    private boolean isOwnerExistsMatchTableName(final SelectStatementContext selectStatementContext, final ColumnProjectionSegment columnProjectionSegment, final String tableName) {
+        if (!columnProjectionSegment.getColumn().getOwner().isPresent()) {
+            return false;
+        }
+        return selectStatementContext.getTablesContext().getAllUniqueTables().stream().anyMatch(table -> tableName.equals(table.getTableName().getIdentifier().getValue()) 
+                && !table.getAlias().isPresent() && columnProjectionSegment.getColumn().getOwner().get().getIdentifier().getValue().equals(tableName));
+    }
+    
+    private boolean isColumnAmbiguous(final SelectStatementContext selectStatementContext, final ColumnProjectionSegment columnProjectionSegment) {
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptProjectionTokenGenerator.java:130]: @LeeGuoPing Is the isColumnAmbiguous method name contrary to the actual logic?

File: shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/table/TablesContext.java [hunk 2]
```diff
@@ -73,6 +73,16 @@ public TablesContext(final Collection<SimpleTableSegment> tableSegments) {
...
     
+    /**
+     * Get all unique table segments.
+     *
+     * @return all unique table segments
+     */
+    public Collection<SimpleTableSegment> getAllUniqueTables() {
+        return uniqueTables.values();
+    }
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/table/TablesContext.java:84]: Please remove this useless blank line?

**Review by strongduanmu — CHANGES_REQUESTED** [review]
@LeeGuoPing Thank you for your effort, this PR is generally great, but there are some details that need to be optimized. In addition, it is best to add encrypt test cases in `shardingsphere-rewrite-test`.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**